mlir/lib/Conversion/ArithToEmitC/ArithToEmitC.cpp

```diff
@@ -739,6 +739,72 @@ class ItoFCastOpConversion : public OpConversionPattern<CastOp> {
   }
 };
 
+class TruncFConversion : public OpConversionPattern<arith::TruncFOp> {
+public:
+  TruncFConversion(const TypeConverter &typeConverter, MLIRContext *context)
+      : OpConversionPattern<arith::TruncFOp>(typeConverter, context) {}
+
+  LogicalResult
+  matchAndRewrite(arith::TruncFOp castOp,
+                  typename arith::TruncFOp::Adaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+    // FIXME Upstream LLVM (commit 77cbc9bf60) brings in a rounding mode
+    // attribute that we need to check. For now, the behavior is the default,
+    // i.e. truncate.
+    Type operandType = adaptor.getIn().getType();
+    if (!emitc::isSupportedFloatType(operandType))
+      return rewriter.notifyMatchFailure(castOp,
+                                         "unsupported cast source type");
+
+    Type dstType = this->getTypeConverter()->convertType(castOp.getType());
+    if (!dstType)
+      return rewriter.notifyMatchFailure(castOp, "type conversion failed");
+
+    if (!castOp.areCastCompatible(operandType, dstType))
+      return rewriter.notifyMatchFailure(castOp, "cast-incompatible types");
+
+    if (!emitc::isSupportedFloatType(dstType) ||
+        (operandType.getIntOrFloatBitWidth() < dstType.getIntOrFloatBitWidth()))
```

Comment:
Maybe move the `!emitc::isSupportedFloatType(dstType)` before the cast compatibility check (with `unsupported cast destination type`)?

Reply:
(same for extf)

Reply:
`areCastCompatible` indeed does this check. I failed to realize that because I can pass in the converted types to this checker, I don't need to perform this check again. Thanks!